File: Data Structures & Algorithms/buy-and-sell-crypto/submission-3.java
class Solution {
    public int maxProfit(int[] prices) {
        int left = 0;
        int right = 0;
        int max = 0;
        

        while(left < prices.length){
            while(right < prices.length && prices[right] >= prices[left]){
                int diff = prices[right] - prices[left];
            
                /*System.out.println("L " + prices[left]);
                System.out.println("R "+prices[right]);

                System.out.println("_____");*/
                if(diff > max){
                    max = diff;
                }

                right++;
            }
            left++;
        }
        
        return max;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.js
const getMostFreq = (map) =>{
    let max = 0
    for (let x of map.values()) {
        if(x > max){
            max = x
        }
    }
    return max
}



class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        let mem = new Map()

        let left = 0
        let right = 0 
        let count = 0
        let max = 0

        while(left < s.length){
            while(right < s.length){
                if(mem.has(s[right])){
                    mem.set(s[right], mem.get(s[right]) +1)
                }else{
                    mem.set(s[right], 1)
                }
                count++
                right++
                let mostFreq = getMostFreq(mem)
                let validCheck =  count - mostFreq
                
                if(validCheck <= k){
                    console.log(count, mem, validCheck, true)
                     if(count > max){
                        max = count
                     }
                }else{
                    console.log(count, mem, validCheck, false)
                    break;
                    
                }
                
                
                
            }
            console.log(count,' move left and remove from mem and count--')
            mem.set(s[left], mem.get(s[left])-1)
            count--
            left++
        }
        return max
    }
}
